src/main/java/net/jodah/failsafe/RetryPolicy.java

```diff
@@ -420,6 +470,37 @@ public RetryPolicy withDelay(long delay, TimeUnit timeUnit) {
     return this;
   }
 
+  /**
+   * Sets the function that determines the next delay before retrying.
+   * @param delayFunction the function to use to compute the delay before a next attempt
+   * @throws NullPointerException if {@code delayFunction} is null
+   * @throws IllegalStateException if backoff delays have already been set
+   */
+  public RetryPolicy withDelay(DelayFunction<Object, Throwable> delayFunction) {
+      return withDelay(delayFunction, Object.class, Throwable.class);
+  }
+
+  /**
+   * Sets the function that determines the next delay before retrying.
+   * @param delayFunction the function to use to compute the delay before a next attempt
+   * @param resultType the type of result from the previous attempt expected by the delay function
+   * @param failureType the type of failure from the previous attempt expected by the delay function
+   * @throws NullPointerException if {@code delayFunction} is null, {@code resultType} is null, or
+   *     {@code failureType} is null
+   * @throws IllegalStateException if backoff delays have already been set
+   */
+  public <R, F extends Throwable> RetryPolicy withDelay(DelayFunction<R, F> delayFunction,
```

Comment:
Coming back to this PR after too-long (apologies) and I like it overall except for this method signature, which seems a bit too prescriptive for certain cases and awkward for others, ex:

```java
rp.withDelay(delayFn, Object.class, TooManyRequestsException.class);
```

Foremost, I'm wondering about the use case for this, where we'd care about the result and exception types together. Most likely we only care about one or the other. So with that in mind, along with consistency with the rest of the API, how about:

```java
rp.withDelayOn(delayFunction, TooManyRequestsException.class); // delay on specific failure

// or

rp.withDelayWhen(delayFunction, 500); // delay on specific result
```

...where the use of the "On" and "When" wording indicates that the delay is conditional. Of course all of these are just convenience methods since `withDelay(DelayFunction)` can always be used by itself.

Thoughts?

As part of this, we also may consider whether to expose the delay result and failure type or if we should just add a `canApplyDelayFunction` to `RetryPolicy` which, similar to `canRetry` and `canAbort` would evaluate the conditionals internally for some result/failure.

Reply:
Looks like you went ahead and merged. I haven't had a chance to respond to your last comment, though.

Reply:
No worries. I wasn't sure if you were tied up (which I can sympathize with) so I decided to just merge the PR as is and make the tweaks I had in mind above afterwards. Feel free to share your thoughts on that whenever you can.

Reply:
Cool. #128 was just me testing something against Travis, but the commits for this are already in master.